backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/LlvmDeclarations.kt

```diff
@@ -262,35 +263,113 @@ private class DeclarationsGeneratorVisitor(override val context: Context) :
-
-        val instanceShadowFieldRef =
-                if (threadLocal) null
-                else {
-                    val shadowSymbolName = if (isExported) {
-                        irClass.objectInstanceShadowFieldSymbolName
-                    } else {
-                        "kshadowobjref:" + qualifyInternalName(irClass)
+        val addressGetterName = if (isExported) {
+            irClass.objectInstanceAddressSymbolName
+        } else {
+            null
+        }
+
+        // If object is imported - access it via getter functions.
+        if (isExternal(irClass)) {
+            assert(valueGetterName != null)
+            assert(addressGetterName != null)
+            val valueGetterFunction = LLVMGetNamedFunction(context.llvmModule, valueGetterName)!!
+            val addressGetterFunction = LLVMGetNamedFunction(context.llvmModule, addressGetterName)!!
+            val instanceGetter = object: ValueAccess() {
+                override fun getValue(generationContext: FunctionGenerationContext?,
+                                      exceptionHandler: ExceptionHandler,
+                                      startLocation: LocationInfo?, endLocation: LocationInfo?): LLVMValueRef {
+                    with(generationContext!!) {
+                        return call(valueGetterFunction, listOf(), exceptionHandler = exceptionHandler)
+                    }
+                }
+            }
+            val addressGetter = object: AddressAccess() {
+                override fun getAddress(generationContext: FunctionGenerationContext?): LLVMValueRef {
+                    with(generationContext!!) {
+                        return call(addressGetterFunction, listOf())
                     }
-                    addGlobal(shadowSymbolName, getLLVMType(irClass.defaultType), isExported = isExported, threadLocal = true)
                 }
+            }
+            return SingletonLlvmDeclarations(instanceGetter, addressGetter)
+        }
 
-        instanceShadowFieldRef?.let { LLVMSetInitializer(it, kNullObjHeaderPtr) }
+        val threadLocal = irClass.storageKind(context) == ObjectStorageKind.THREAD_LOCAL
+        val symbolName = "kobjref:" + qualifyInternalName(irClass)
+        val instanceAddress = addKotlinGlobal(symbolName, getLLVMType(irClass.defaultType), threadLocal = threadLocal)
+
+        val instanceShadowAddress = if (threadLocal) null else {
+            val shadowSymbolName = "kshadowobjref:" + qualifyInternalName(irClass)
+            addKotlinGlobal(shadowSymbolName, getLLVMType(irClass.defaultType), threadLocal = true)
+        }
+
+        val instanceGetter = object: ValueAccess() {
+            override fun prepare(codegen: CodeGenerator) {
+                if (valueGetterName != null) {
+                    generateFunction(codegen, LLVMFunctionType(kObjHeaderPtr, cValuesOf(kObjHeaderPtrPtr), 1, 0)!!,
```

Comment:
Generally `LlvmDeclarations` machinery isn't supposed to generate any _definitions_ (including function bodies). So it makes sense to move all these new instruction generators to `IrToBitcode` to keep it consistent with the existing code.